psi4/driver/p4util/solvers.py

```diff
@@ -308,3 +308,450 @@ def extrapolate(self, out=None):
+            iter_info['val'][k] = w[k]
+            if (iter_info['res_norm'][k] < r_tol) and (iter_info['delta_val'][k] < e_tol):
+                continue
+            else:
+                iter_info['done'] = False
+                if iter_info['collapse']:
+                    continue
+                else:
+                    X_new.append(engine.precondition_residual(Rv, shift=w[k]))
+
+        _diag_print_info(print_name, iter_info, verbose)
+        stats.append(iter_info.copy())
+        if iter_info['done']:
+            _diag_print_converged(print_name, stats, w, rvec=V, verbose=verbose)
+            return w, V, stats
+        if iter_info['collapse']:
+            X = []
+            X_new = V
+
+    # If we get down here  we have exceeded max iterations without convergence, return none + stats
+    return None, None, stats
+
+
+def hamiltonian_solver(engine,
+                       guess,
+                       e_tol=1.0E-6,
+                       r_tol=1.0E-8,
+                       nroot=1,
+                       max_vecs_per_root=20,
+                       maxiter=100,
+                       verbose=1):
+    """
+    Finds the smallest eigenvalues and associated right and left hand eigenvectors of a large real Hamiltonian eigenvalue problem
+    emulated through an engine.
+
+    Similar to the davidson algorithm but with special considerations that preserve the structure of the problem and lead to faster
+    convergence.
+
+    Parameters
+    -----------
+    engine : object
+       The engine drive all operations involving data structures that have one "large" dimension. The methods it is required
+       to provide and their signatures are detailed below.
+    guess : list {engine dependent}
+       At least `nroot` initial expansion vectors
+    e_tol : float
+        Convergence tolerance for eigenvalues
+    r_tol : float
+        Convergence tolerance for residual vectors
+    nroot : int
+        Number of roots desired
+    maxiter : int
+        The maximum number of iterations
+    verbose : int
+        The amount of logging info to print (0 -> none, 1 -> some, >1 -> everything)
+
+    Requirements of engine
+    ----------------------
+    orthogonalize_subspace(old_X, new_X) --> X, nx:
+       old_X : list of `vectors`
+           This set may be empty, if not, the vectors are mutually orthogonal.
+       new_x : list of `trial vectors`
+           The normalization or orthogonality of these vectors can't be assumed
+       X : list of `vectors`
+           A set of ortho-normal vectors constructed from old_X and new_X, need not have dimension = dim(old_X) + dim(new_X)
+       nl : The number of vectors in the set.
+
+    compute_products(X) -> Px, Mx:
+       X : list of `vectors`
+           As returned from `orthogonalize_subspace`
+       Px : list of `vectors`
+           The product :math:`(A+B)X_{i}` for each `X_{i}` in `X`, in that order.
```

Comment:
I guess I made that up? I can change this to H1/H2 if that is the standard language. I wasn't aware of that while I was writing the code and In my notes I used P/M so that carried over to the code.